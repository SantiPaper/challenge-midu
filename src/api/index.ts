import books from "./books.json";
import type { Book } from "../types/books";

export const getBooks = (): Promise<Book[]> => {
  return Promise.resolve(books.library.map((book) => book.book));
};
