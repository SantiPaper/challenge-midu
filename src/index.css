* {
  padding: 0;
  margin: 0;
  box-sizing: border-box;
}

html {
  font-size: 10px;
  --primary-200: #e7cc90;
  --primary-400: #feb510;
  --secondary-400: #788ee5;
  --secondary-200: #9ca9df;
  --black: #090f15;

  --font-xl: 96px;
  --font-l: 64px;
  --font-m: 36px;
  --font-s: 24px;
  --font-xs: 20px;
  --font-xxs: 14px;

  --font-heading: "Yeseva One", cursive;
  --font-body: "Inter", Helvetica, Arial, Sans-serif;
}

body {
  background-color: var(--primary-400);
  font-family: var(--font-body);
}
