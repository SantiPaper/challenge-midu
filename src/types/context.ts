import { Book } from "./books";

export type Context = {
  state: State;
  addBooks: (books: Book[]) => void;
  getByGenre: (genre: Genre) => void;
  getBySearch: (search: string) => void;
  getByPages: (pages: number) => void;
  addToLectureList: (book: Book) => void;
  removeFromLectureList: (book: Book) => void;
  /* mostrarModal: (book: Book) => void; */
  setDetail: (book: Book) => void;
};

export type Genre = "Ciencia ficción" | "Fantasía" | "Terror" | "Zombies" | "";

export type State = {
  books: Book[];
  readingList: Book[];
  filteredBooks: Book[];
  detail: Book[];
};

export type Reducer = (state: State, action: Action) => State;

export type Action =
  | {
      type: "ADD_BOOKS";
      payload: Book[];
    }
  | {
      type: "GET_BY_GENRE";
      payload: Book[];
    }
  | {
      type: "GET_BY_SEARCH";
      payload: Book[];
    }
  | {
      type: "GET_BY_PAGES";
      payload: Book[];
    }
  | {
      type: "ADD_TO_LECTURE_LIST";
      payload: Book;
    }
  | {
      type: "REMOVE_FROM_LECTURE_LIST";
      payload: string;
    }
  | {
      type: "SET_DETAIL";
      payload: Book;
    };
