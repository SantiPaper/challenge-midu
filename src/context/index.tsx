import { createContext, PropsWithChildren, useReducer } from "react";
import type { Context, Genre } from "../types/context";
import { reducer, initialState } from "./reducer";
import { Book } from "../types/books";

export const BooksContext = createContext<Context | null>(null);

export const Provider = ({ children }: PropsWithChildren) => {
  const [state, dispatch] = useReducer(reducer, initialState);

  const addBooks = (books: Book[]) => {
    dispatch({
      type: "ADD_BOOKS",
      payload: books,
    });
  };

  const getByGenre = (genre: Genre) => {
    dispatch({
      type: "GET_BY_GENRE",
      payload: genre
        ? state.books.filter((book) => book.genre === genre)
        : state.books,
    });
  };

  const getBySearch = (search: string) => {
    dispatch({
      type: "GET_BY_SEARCH",
      payload: search
        ? state.books.filter((book) =>
            book.title.toLowerCase().includes(search.toLowerCase())
          )
        : state.books,
    });
  };

  const getByPages = (pages: number) => {
    dispatch({
      type: "GET_BY_SEARCH",
      payload: state.books.filter((book) => book.pages > pages),
    });
  };

  const addToLectureList = (book: Book) => {
    dispatch({
      type: "ADD_TO_LECTURE_LIST",
      payload: book,
    });
  };

  const removeFromLectureList = (book: Book) => {
    dispatch({
      type: "REMOVE_FROM_LECTURE_LIST",
      payload: book.ISBN,
    });
  };

  const setDetail = (book: Book) => {
    dispatch({
      type: "SET_DETAIL",
      payload: book,
    });
  };

  return (
    <BooksContext.Provider
      value={{
        state,
        addBooks,
        getByGenre,
        getBySearch,
        getByPages,
        addToLectureList,
        removeFromLectureList,
        /* showDetail,
        detail,
        mostrarModal, */
        setDetail,
      }}
    >
      {children}
    </BooksContext.Provider>
  );
};
