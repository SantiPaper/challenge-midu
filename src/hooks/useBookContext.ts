import { useContext } from "react";
import { BooksContext } from "../context";

export const useBookContext = () => {
  const value = useContext(BooksContext);
  if (!value) {
    throw new Error("useBookContext tiene que usarse dentro del provider");
  }
  return value;
};
