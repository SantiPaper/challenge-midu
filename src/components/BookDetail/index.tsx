import { StyleDetail } from "./style";
import { useBookContext } from "../../hooks/useBookContext";
import { useEffect } from "react";
import type { Book } from "../../types/books";

type Props = {
  onClose: () => void;
};

export const BookDetail = ({ onClose }: Props) => {
  const { state, addToLectureList, removeFromLectureList } = useBookContext();

  const escapeListener = (e: Event) => {
    if ((e as unknown as KeyboardEvent).key === "Escape") {
      onClose();
    }
  };

  const mouseListener = (e: Event) => {
    if ((e.target as HTMLDivElement).id === "close-modal") {
      onClose();
    }
  };

  useEffect(() => {
    window.addEventListener("keydown", escapeListener);
    window.addEventListener("click", mouseListener);

    return () => {
      window.removeEventListener("keydown", escapeListener);
      window.removeEventListener("click", mouseListener);
    };
  }, []);

  const onClick = (book: Book) => {
    if (state.readingList.includes(book)) {
      removeFromLectureList(book);
    } else addToLectureList(book);
  };

  return (
    <StyleDetail id="close-modal">
      <div className="modal">
        {state.detail?.map((book) => (
          <article className="card" key={book.ISBN}>
            <img
              className="card__img"
              width={300}
              src={book.cover}
              alt={book.title}
            />
            <div className="container-info">
              <h2 className="card__title">{book.title}</h2>
              <p>By {book.author.name}</p>
              <p>Genero/{book.genre}</p>
              <p className="card__synopsis">{book.synopsis}</p>
              <p> {book.pages} Paginas</p>
              <p>Lanzado en {book.year}</p>
              <button
                className="card__add-lecture"
                onClick={() => onClick(state.detail[0])}
              >
                {state.readingList.includes(state.detail[0])
                  ? "Quitar de lista de lectura"
                  : "Agregar a lista de lectura"}
              </button>
            </div>
          </article>
        ))}
        <button className="close-modal" onClick={() => onClose()}>
          X
        </button>
      </div>
    </StyleDetail>
  );
};
