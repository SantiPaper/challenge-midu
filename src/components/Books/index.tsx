import { useBookContext } from "../../hooks/useBookContext";
import { StyleBookSection } from "./style";

export const Books = () => {
  const { state } = useBookContext();

  return (
    <StyleBookSection>
      {state.books.map((book) => (
        <article className="book__container" key={book.ISBN}>
          <img className="book__image" src={book.cover} width={120} alt="" />
          <div className="book__card">
            <h2 className="book__title">
              {book.title.length < 25
                ? book.title
                : book.title.slice(0, 25).concat("(...)")}
            </h2>
            <div className="book__container-info">
              <p>{book.author.name}</p>
              <p>{book.year}</p>
            </div>
            <div className="book__container-buttons">
              <button>Agregar a lista de lectura</button>
              <button>Ver mas</button>
            </div>
          </div>
        </article>
      ))}
    </StyleBookSection>
  );
};
