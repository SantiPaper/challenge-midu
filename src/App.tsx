import { StyledContainer } from "./style";
import { Header } from "./components/Header";
import { Main } from "./components/Main";

function App() {
  return (
    <>
      <StyledContainer>
        <Header />
        <Main />
      </StyledContainer>
    </>
  );
}

export default App;
