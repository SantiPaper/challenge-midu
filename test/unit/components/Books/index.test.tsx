import { render } from "@testing-library/react";
import { Books } from "../../../../src/components/Books";
import { Provider } from "../../../../src/context";
import * as contextModule from "../../../../src/hooks/useBookContext";
import { mockArr } from "../../mock";

describe("Componente books", () => {
  vi.spyOn(contextModule, "useBookContext").mockReturnValue({
    state: { filteredBooks: mockArr },
  });
  test("Renderizan todas las cards", async () => {
    const { container } = render(
      <Provider>
        <Books />
      </Provider>
    );

    expect(container).toMatchSnapshot();
  });
});
