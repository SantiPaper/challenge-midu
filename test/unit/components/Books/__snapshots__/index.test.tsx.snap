// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Componente books > Renderizan todas las cards 1`] = `
<div>
  <section
    class="sc-bcPKhP cnNqQw"
  >
    <article
      class="book__container"
    >
      <img
        alt=""
        class="book__image"
        src="https://images-na.ssl-images-amazon.com/images/S/compressed.photo.goodreads.com/books/1273763400i/8189620.jpg"
        width="120"
      />
      <div
        class="book__card"
      >
        <h2
          class="book__title"
        >
          Juego de Tronos
        </h2>
        <div
          class="book__container-info"
        >
          <p>
            George R. R. Martin
          </p>
          <p>
            1996
          </p>
        </div>
        <div
          class="book__container-buttons"
        >
          <button
            class="book__button-ls"
          >
            Agregar a lista de lectura
          </button>
          <button
            class="book__button-wm"
          >
            Ver más
          </button>
        </div>
      </div>
    </article>
    <article
      class="book__container"
    >
      <img
        alt=""
        class="book__image"
        src="https://images-na.ssl-images-amazon.com/images/S/compressed.photo.goodreads.com/books/1550337333i/15868.jpg"
        width="120"
      />
      <div
        class="book__card"
      >
        <h2
          class="book__title"
        >
          Harry Potter y la piedra filosofal
        </h2>
        <div
          class="book__container-info"
        >
          <p>
            J.K. Rowling
          </p>
          <p>
            1997
          </p>
        </div>
        <div
          class="book__container-buttons"
        >
          <button
            class="book__button-ls"
          >
            Agregar a lista de lectura
          </button>
          <button
            class="book__button-wm"
          >
            Ver más
          </button>
        </div>
      </div>
    </article>
  </section>
</div>
`;
